src/replay/consensus/core.zig

```diff
@@ -1,235 +1,661 @@
+            // from the root slot + 1
+            root_slot +| 1,
         );
-        for (params.progress_map.map.keys(), params.progress_map.map.values()) |slot, prog| {
-            if (prog.fork_stats.duplicate_confirmed_hash != null) {
-                continue;
+        defer duplicate_slots.deinit();
+
+        while (try duplicate_slots.nextKey()) |slot| {
+            const ref = slot_tracker.get(slot) orelse continue;
+            try heaviest_subtree_fork_choice.markForkInvalidCandidate(&.{
+                .slot = slot,
+                .hash = ref.state.hash.readCopy().?,
+            });
+        }
+
+        return heaviest_subtree_fork_choice;
+    }
+
+    /// Run all phases of consensus:
+    /// - process replay results
+    /// - edge cases
+    /// - actual consensus protocol.
+    pub fn process(
+        self: *TowerConsensus,
+        allocator: Allocator,
+        slot_tracker_rw: *RwMux(SlotTracker),
+        epoch_tracker_rw: *RwMux(EpochTracker),
+        progress_map: *ProgressMap,
+        results: []const ReplayResult,
+    ) !void {
+        var arena_state = self.arena_state.promote(allocator);
+        defer {
+            _ = arena_state.reset(.retain_capacity);
+            self.arena_state = arena_state.state;
+        }
+        const arena = arena_state.allocator();
+
+        { // Process replay results
+            const slot_tracker, var lock = slot_tracker_rw.readWithLock();
+            defer lock.unlock();
+            for (results) |r| try self.processResult(allocator, progress_map, slot_tracker, r);
+        }
+
+        // Process edge cases and prepare ancestors/descendants
+        const ancestors, const descendants = edge_cases_and_ancestors_descendants: {
+            const slot_tracker, var slot_tracker_lg = slot_tracker_rw.readWithLock();
+            defer slot_tracker_lg.unlock();
+
+            _ = try replay.consensus.edge_cases.processEdgeCases(allocator, .from(self.logger), .{
+                .my_pubkey = self.my_identity,
+                .tpu_has_bank = false,
+                .fork_choice = &self.fork_choice,
+                .ledger = self.ledger_writer,
+                .slot_tracker = slot_tracker,
+                .progress = progress_map,
+                .latest_validator_votes = &self.latest_validator_votes,
+                .slot_data = &self.slot_data,
+                .senders = self.senders,
+                .receivers = self.receivers,
+            });
+
+            const SlotSet = SortedSetUnmanaged(Slot);
+
+            // arena-allocated
+            var ancestors: std.AutoArrayHashMapUnmanaged(Slot, Ancestors) = .empty;
+            var descendants: std.AutoArrayHashMapUnmanaged(Slot, SlotSet) = .empty;
+            for (slot_tracker.slots.keys(), slot_tracker.slots.values()) |slot, info| {
+                const slot_ancestors = &info.constants.ancestors.ancestors;
+                const ancestor_gop = try ancestors.getOrPutValue(arena, slot, .EMPTY);
```

Comment:
Seems a bit counter-intuitive that `ancestors` is part of `constants`, but is being mutated here. Should it maybe instead be part of SlotState? If not, an explanation turned into a comment explaining why would be appreciated.

Reply:
This is the map containing a collection of instances of `Ancestors`. It's not the individual instance of `Ancestors` found in slot constants.